Change `st.text` font and overflow styles

sfc-gh-pchiu opened the pull request:
## Describe your changes
### Context
Today, `st.text` renders the text in monospace and makes it scrollable (i.e. no line wrapping):
![image](https://github.com/user-attachments/assets/1cbe7ea7-128f-4196-8992-8d24618d0f6c)

The look differs from `st.markdown` and `st.write`, and the behavior by itself is a bit inconsistent and not very useful.

### In this PR
1. Use normal font for `st.text` instead of monospace font
2. Switch to wrapping long `st.text` from using scroll bar when overflowed
3. Update e2e test cases

## Testing Plan
- [x] E2E Tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
change st.text font to be not monospaced

sfc-gh-pchiu commented on e2e_playwright/st_text.py:
```diff
@@ -15,12 +15,5 @@
- \_^_/
-"""
-)
 st.text("_This text is **awesome**!_")
```
removing this since we're not using monospace font :'(

sfc-gh-pchiu committed:
update snapshots

sfc-gh-pchiu committed:
fix e2e test after removing ascii art

sfc-gh-pchiu committed:
snapshots

sfc-gh-pchiu committed:
update text font, overflow style

raethlein commented on e2e_playwright/st_text_test.py:
```diff
@@ -22,26 +22,33 @@ def test_st_text_shows_correct_text(app: Page):
     expect_help_tooltip(app, text_with_help, "This is a help tooltip!")
 
 
+def test_multiline_text(app: Page):
```
Great that you have added a test for the multiline behavior! One thought: does this really ensure that `\n` is rendered as a newline? Could the text also simply contain the `\n` characters and the test would still pass? Perhaps not if `\n` is always treated by the browser as a newline for text, but maybe worth double-checking that the test is indeed ensuring what we believe it does 🙂

sfc-gh-pchiu replied:
It doesn't as I was testing that this case indeed exists in `st_text.py` (which I used for manual testing.)

raethlein commented:
Based on the spec, the `st.text` command's docstring should be updated as well:
> Update the docstring of st.text to remove “fixed-width”. Just say that it renders normal text without markdown capabilities.

sfc-gh-pchiu committed:
correct font size

sfc-gh-pchiu committed:
test multiline

sfc-gh-pchiu reviewed (commented):
Cool! I'll seek guidance from @sfc-gh-dmatthews for the docstring

sfc-gh-pchiu commented on e2e_playwright/st_text_test.py:
```diff
@@ -22,26 +22,33 @@ def test_st_text_shows_correct_text(app: Page):
+    """Test that multiline text is displayed correctly."""
+    multiline_text = app.get_by_test_id("stText").nth(3).locator("div")
+    bounding_box = multiline_text.bounding_box()
+    assert bounding_box["height"] > bounding_box["width"]
```
@raethlein I didn't find a good way to validate if we actually render newlines here. Suggestions were like to look for line break characters.
I ended up checking the dimensions here, given that our `st.text` is standardized in font family and size. (A future edge case would be that we use some ultra-wide font for `st.text`)

raethlein replied:
I like the idea 🙂 An alternative way might be to check whether there is an actual character sequence `\n` in the text itself. Though if this is not possible or too tricky, it should be fine to leave a comment to this approach and move on 👍

sfc-gh-pchiu replied:
Added
1. `expect(multiline_text).not_to_contain_text("\\n")` to multi-line test to ensure we don't actually display the `\n` character sequence
2. single line test *with* literal `\n` character sequence and ensure it doesn't do line break, i.e. `expect(singleline_text).not_to_contain_text(repr("\n"))` being true

sfc-gh-pchiu committed:
snapshots

sfc-gh-pchiu committed:
snapshots

sfc-gh-pchiu committed:
snapshots

sfc-gh-pchiu committed:
snapshots

sfc-gh-dmatthews committed:
Update docstrings for `st.text` changes (#9592)

## Describe your changes
Docs for `st.text` changes.

---

**Contribution License Agreement**
By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
snapshots

sfc-gh-pchiu committed:
snapshots

sfc-gh-pchiu committed:
more.."

sfc-gh-pchiu committed:
snapshots

sfc-gh-pchiu committed:
..

sfc-gh-pchiu committed:
...

raethlein commented on lib/streamlit/elements/code.py:
```diff
@@ -40,32 +40,57 @@ def code(
         ----------
         body : str
-            The string to display as code.
+            The string to display as code or monospace text.
```
Just for clarity, these doc string changes probably come after a discussion with Debbie and are not directly related to the `st.text` PR, right?

sfc-gh-pchiu replied:
Yes! These were updated in #9592 (I may prefer to split the non-`st.text` changes out. But I also see how this change would affect other APIs as an entirety)

raethlein replied:
yeah its fine to leave them in here 🙂 thanks for verifying!

raethlein reviewed (approved):
LGTM!

sfc-gh-pchiu committed:
test

sfc-gh-pchiu committed:
test

sfc-gh-pchiu committed:
add back escape char

sfc-gh-pchiu committed:
remove
